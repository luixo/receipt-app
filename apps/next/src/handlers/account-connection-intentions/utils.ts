import * as trpc from "@trpc/server";

import { Database } from "next-app/db";
import { ACCOUNT_CONNECTIONS_INTENTIONS } from "next-app/db/consts";
import { AccountsId, UsersId } from "next-app/db/models";
import { getUserById } from "next-app/handlers/users/utils";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: string,
	asUserId: UsersId
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select("id")
		.where("email", "=", toEmail)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Account with email ${toEmail} does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId)
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new trpc.TRPCError({
			code: "CONFLICT",
			message: `Account with email ${toEmail} is already connected to user ${connectedUser.name}.`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where("accountId", "=", targetAccount.id)
		.where("targetAccountId", "=", fromAccountId)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where("ownerAccountId", "=", targetAccount.id)
				.where("id", "=", viceVersaIntention.userId)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where("ownerAccountId", "=", fromAccountId)
				.where("id", "=", asUserId)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where("accountId", "=", targetAccount.id)
				.where("targetAccountId", "=", fromAccountId)
				.executeTakeFirst();
		});
		return {
			id: targetAccount.id,
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				created: new Date(),
			})
			.executeTakeFirst();
		return {
			id: targetAccount.id,
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where("accountId", "=", fromAccountId)
				.where("targetAccountId", "=", targetAccount.id)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await getUserById(
				database,
				existingIntention.userId,
				["name"]
			);
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to ${toEmail} as user ${
					existingUser!.name
				}.`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new trpc.TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user ${user.name}.`,
			});
		}
		throw e;
	}
};

export const removeIntention = async (
	database: Database,
	intention:
		| {
				targetAccountId: AccountsId;
				accountId: AccountsId;
		  }
		| undefined,
	intentionType: string
) => {
	if (!intention) {
		throw new trpc.TRPCError({
			code: "NOT_FOUND",
			message: `Intention ${intentionType} does not exist.`,
		});
	}
	await database
		.deleteFrom("accountConnectionsIntentions")
		.where("accountId", "=", intention.accountId)
		.where("targetAccountId", "=", intention.targetAccountId)
		.execute();
};
