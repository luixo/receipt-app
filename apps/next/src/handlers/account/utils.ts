import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { AccountsId } from "../../db/models";

export const getAccountById = <SE extends ReceiptsSelectExpression<"accounts">>(
	database: Database,
	id: AccountsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};

export const getAccountByEmail = <
	SE extends ReceiptsSelectExpression<"accounts">
>(
	database: Database,
	email: string,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "accounts", SE> | undefined> => {
	return database
		.selectFrom("accounts")
		.select(selectExpression)
		.where("email", "=", email)
		.executeTakeFirst();
};
