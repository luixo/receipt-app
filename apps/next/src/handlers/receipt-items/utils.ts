import { Selection } from "kysely";
import { Database, ReceiptsSelectExpression, ReceiptsDatabase } from "../../db";
import { ReceiptItemsId } from "../../db/models";

export const getReceiptItemById = <
	SE extends ReceiptsSelectExpression<"receipt_items">
>(
	database: Database,
	id: ReceiptItemsId,
	selectExpression: SE[]
): Promise<Selection<ReceiptsDatabase, "receipt_items", SE> | undefined> => {
	return database
		.selectFrom("receipt_items")
		.select(selectExpression)
		.where("id", "=", id)
		.executeTakeFirst();
};
