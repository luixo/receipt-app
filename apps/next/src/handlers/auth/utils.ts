import * as trpc from "@trpc/server";

import { DAY } from "app/utils/time";
import type { Database } from "next-app/db";
import type { AccountsId, SessionsSessionId } from "next-app/db/models";
import { generateConfirmEmailEmail } from "next-app/email/utils";
import { getUuid } from "next-app/utils/crypto";
import { getEmailClient, isEmailServiceActive } from "next-app/utils/email";

// How long a session should last
const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long before the session expires should we update the expiration date
const SESSION_SHOULD_UPDATE_DURATION = 28 * DAY;

const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);
export const shouldUpdateExpirationDate = (expirationTimestamp: Date) =>
	expirationTimestamp.valueOf() - Date.now() < SESSION_SHOULD_UPDATE_DURATION;

export const createAuthorizationSession = async (
	database: Database,
	accountId: AccountsId,
) => {
	const uuid = getUuid();
	const expirationDate = getExpirationDate();
	await database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const updateAuthorizationSession = async (
	database: Database,
	authToken: SessionsSessionId,
) => {
	database
		.updateTable("sessions")
		.set({ expirationTimestamp: getExpirationDate() })
		.where("sessionId", "=", authToken)
		.executeTakeFirst();
};

export const removeAuthorizationSession = async (
	database: Database,
	sessionId: SessionsSessionId,
) => {
	await database
		.deleteFrom("sessions")
		.where("sessionId", "=", sessionId)
		.executeTakeFirst();
};

export const sendVerificationEmail = async (email: string, token: string) => {
	if (!isEmailServiceActive()) {
		throw new trpc.TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient().send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(token),
		});
	} catch (e) {
		throw new trpc.TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
