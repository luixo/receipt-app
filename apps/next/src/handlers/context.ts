import * as trpcNext from "@trpc/server/adapters/next";
import { NextApiRequest, NextApiResponse } from "next";
import { Logger, logger } from "../utils/logger";

export type Context = {
	req: NextApiRequest;
	res: NextApiResponse;
	logger: Logger;
	debug: boolean;
};

export const createContext = (
	opts: trpcNext.CreateNextContextOptions
): Context => ({
	req: opts.req,
	res: opts.res,
	logger,
	debug: Boolean(opts.req.headers.debug),
});
