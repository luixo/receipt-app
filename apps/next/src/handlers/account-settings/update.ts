import { z } from "zod";

import type { SimpleUpdateObject } from "next-app/db/types";
import { authProcedure } from "next-app/handlers/trpc";

import { DEFAULT_ACCOUNT_SETTINGS } from "./get";

type SettingsUpdateObject = SimpleUpdateObject<"accountSettings">;

export const procedure = authProcedure
	.input(
		z.discriminatedUnion("type", [
			z.strictObject({
				type: z.literal("autoAcceptDebts"),
				value: z.boolean(),
			}),
		]),
	)
	.mutation(async ({ ctx, input }) => {
		const { database } = ctx;
		const updateObject: SettingsUpdateObject = {};
		switch (input.type) {
			case "autoAcceptDebts":
				updateObject.autoAcceptDebts = input.value;
		}
		const existingSettings = await database
			.selectFrom("accountSettings")
			.where("accountSettings.accountId", "=", ctx.auth.accountId)
			.executeTakeFirst();
		if (!existingSettings) {
			await database
				.insertInto("accountSettings")
				.values({
					accountId: ctx.auth.accountId,
					...DEFAULT_ACCOUNT_SETTINGS,
					...updateObject,
				})
				.executeTakeFirst();
		} else {
			await database
				.updateTable("accountSettings")
				.set(updateObject)
				.where("accountSettings.accountId", "=", ctx.auth.accountId)
				.executeTakeFirst();
		}
	});
