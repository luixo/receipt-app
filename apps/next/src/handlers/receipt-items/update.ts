import * as trpc from "@trpc/server";
import { MutationObject } from "kysely";
import { z } from "zod";

import { ReceiptsDatabase, getDatabase } from "../../db";
import { AuthorizedContext } from "../context";
import { getReceiptById, getAccessRole } from "../receipts/utils";
import { getReceiptItemById } from "./utils";
import { flavored } from "../zod";
import { ReceiptItemsId } from "../../db/models";

export const router = trpc.router<AuthorizedContext>().mutation("update", {
	input: z.strictObject({
		id: z.string().uuid().refine<ReceiptItemsId>(flavored),
		update: z.discriminatedUnion("type", [
			z.strictObject({
				type: z.literal("name"),
				name: z.string().min(2).max(255),
			}),
			z.strictObject({ type: z.literal("price"), price: z.number().gt(0) }),
			z.strictObject({
				type: z.literal("quantity"),
				quantity: z.number().gt(0),
			}),
			z.strictObject({ type: z.literal("locked"), locked: z.boolean() }),
		]),
	}),
	resolve: async ({ input, ctx }) => {
		const database = getDatabase(ctx);
		const receiptItem = await getReceiptItemById(database, input.id, [
			"receiptId",
		]);
		if (!receiptItem) {
			throw new trpc.TRPCError({
				code: "PRECONDITION_FAILED",
				message: `No receipt item found by id ${input.id}`,
			});
		}
		const receipt = await getReceiptById(database, receiptItem.receiptId, [
			"id",
			"ownerAccountId",
		]);
		if (!receipt) {
			throw new trpc.TRPCError({
				code: "PRECONDITION_FAILED",
				message: `No receipt found by id ${input.id}`,
			});
		}
		const accessRole = await getAccessRole(
			database,
			receipt,
			ctx.auth.accountId
		);
		if (accessRole !== "owner" && accessRole !== "editor") {
			throw new trpc.TRPCError({
				code: "UNAUTHORIZED",
				message: `Receipt ${receipt.id} is not allowed to be modified by ${
					ctx.auth.accountId
				} with role ${accessRole || "nobody"}`,
			});
		}
		let setObject: MutationObject<
			ReceiptsDatabase,
			"receiptItems",
			"receiptItems"
		> = {};
		switch (input.update.type) {
			case "name":
				setObject = { name: input.update.name };
				break;
			case "price":
				setObject = { price: input.update.price.toString() };
				break;
			case "quantity":
				setObject = { quantity: input.update.quantity.toString() };
				break;
			case "locked":
				setObject = { locked: input.update.locked };
				break;
		}
		await getDatabase(ctx)
			.updateTable("receiptItems")
			.set(setObject)
			.where("id", "=", input.id)
			.executeTakeFirst();
	},
});
