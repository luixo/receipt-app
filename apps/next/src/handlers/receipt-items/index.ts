import { t } from "next-app/handlers/trpc";

import { procedure as add } from "./add";
import { procedure as get } from "./get";
import { procedure as remove } from "./remove";
import { procedure as update } from "./update";

export const router = t.router({
	get,
	add,
	remove,
	update,
});
