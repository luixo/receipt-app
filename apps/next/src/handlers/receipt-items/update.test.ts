import { faker } from "@faker-js/faker";
import { describe } from "vitest";

import { createAuthContext } from "@tests/backend/utils/context";
import {
	insertAccount,
	insertAccountWithSession,
	insertConnectedUsers,
	insertItemParticipant,
	insertReceipt,
	insertReceiptItem,
	insertReceiptParticipant,
	insertUser,
} from "@tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "@tests/backend/utils/expect";
import type { TestContext } from "@tests/backend/utils/test";
import { test } from "@tests/backend/utils/test";
import type { TRPCMutationInput } from "app/trpc";
import { t } from "next-app/handlers/trpc";

import {
	verifyName,
	verifyPrice,
	verifyQuantity,
	verifyReceiptItemId,
} from "./test.utils";
import { procedure } from "./update";

const router = t.router({ procedure });

const runTests = (
	getUpdate: () => TRPCMutationInput<"receiptItems.update">["update"],
	lockedBefore = false,
) => {
	const runTest = async (ctx: TestContext, type: "own" | "foreign") => {
		const { sessionId, accountId } = await insertAccountWithSession(ctx);
		const { id: receiptId } = await insertReceipt(
			ctx,
			accountId,
			lockedBefore ? { lockedTimestamp: new Date("2020-06-01") } : undefined,
		);
		const { id: userId } = await insertUser(ctx, accountId);
		const { id: receiptItemId } = await insertReceiptItem(ctx, receiptId);
		await insertReceiptParticipant(ctx, receiptId, userId);
		await insertItemParticipant(ctx, receiptItemId, userId);

		// Verify unrelated data doesn't affect the result
		const { id: anotherUserId } = await insertUser(ctx, accountId);
		const { id: anotherReceiptId } = await insertReceipt(ctx, accountId);
		await insertReceiptParticipant(ctx, anotherReceiptId, anotherUserId);
		await insertReceiptItem(ctx, anotherReceiptId);

		const { id: foreignAccountId } = await insertAccount(ctx);
		const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);
		const { id: foreignReceiptId } = await insertReceipt(ctx, foreignAccountId);
		const [{ id: foreignToSelfUserId }] = await insertConnectedUsers(ctx, [
			foreignAccountId,
			accountId,
		]);
		await insertReceiptParticipant(ctx, foreignReceiptId, foreignUserId);
		await insertReceiptParticipant(ctx, foreignReceiptId, foreignToSelfUserId, {
			role: "editor",
		});
		const { id: foreignReceiptItemId } = await insertReceiptItem(
			ctx,
			foreignReceiptId,
		);

		const caller = router.createCaller(createAuthContext(ctx, sessionId));
		await expectDatabaseDiffSnapshot(ctx, () =>
			caller.procedure({
				id: type === "own" ? receiptItemId : foreignReceiptItemId,
				update: getUpdate(),
			}),
		);
	};
	test("own", async ({ ctx }) => {
		await runTest(ctx, "own");
	});
	test("foreign", async ({ ctx }) => {
		await runTest(ctx, "foreign");
	});
};

describe("receiptItems.update", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			router.createCaller(context).procedure({
				id: faker.string.uuid(),
				update: { type: "name", name: faker.lorem.words() },
			}),
		);

		verifyReceiptItemId(
			(context, receiptId) =>
				router.createCaller(context).procedure({
					id: receiptId,
					update: { type: "name", name: faker.lorem.words() },
				}),
			"",
		);

		verifyName(
			(context, name) =>
				router.createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "name", name },
				}),
			"update.",
		);

		verifyPrice(
			(context, price) =>
				router.createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "price", price },
				}),
			"update.",
		);

		verifyQuantity(
			(context, quantity) =>
				router.createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "quantity", quantity },
				}),
			"update.",
		);

		test("receipt item does not exist", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			const { id: receiptId } = await insertReceipt(ctx, accountId);
			await insertReceiptItem(ctx, receiptId);
			const fakeReceiptItemId = faker.string.uuid();
			await expectTRPCError(
				() =>
					caller.procedure({
						id: fakeReceiptItemId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"NOT_FOUND",
				`Receipt item "${fakeReceiptItemId}" is not found.`,
			);
		});

		test("receipt is not owned by an account", async ({ ctx }) => {
			const { sessionId, accountId, account } = await insertAccountWithSession(
				ctx,
			);
			await insertReceipt(ctx, accountId);

			const { id: foreignAccountId } = await insertAccount(ctx);
			const { id: foreignReceiptId } = await insertReceipt(
				ctx,
				foreignAccountId,
			);
			const { id: foreignReceiptItemId } = await insertReceiptItem(
				ctx,
				foreignReceiptId,
			);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						id: foreignReceiptItemId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"FORBIDDEN",
				`Receipt "${foreignReceiptId}" is not allowed to be modified by "${account.email}".`,
			);
		});

		test("receipt role is lower than editor", async ({ ctx }) => {
			const { sessionId, accountId, account } = await insertAccountWithSession(
				ctx,
			);
			await insertReceipt(ctx, accountId);

			const { id: foreignAccountId } = await insertAccount(ctx);
			const [{ id: foreignToSelfUserId }] = await insertConnectedUsers(ctx, [
				foreignAccountId,
				accountId,
			]);
			const { id: foreignReceiptId } = await insertReceipt(
				ctx,
				foreignAccountId,
			);
			await insertReceiptParticipant(
				ctx,
				foreignReceiptId,
				foreignToSelfUserId,
				{ role: "viewer" },
			);
			const { id: foreignReceiptItemId } = await insertReceiptItem(
				ctx,
				foreignReceiptId,
			);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						id: foreignReceiptItemId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"FORBIDDEN",
				`Receipt "${foreignReceiptId}" is not allowed to be modified by "${account.email}" with role "viewer"`,
			);
		});

		test("receipt is locked", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: receiptId } = await insertReceipt(ctx, accountId, {
				lockedTimestamp: new Date(),
			});
			const { id: receiptItemId } = await insertReceiptItem(ctx, receiptId);

			const caller = router.createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						id: receiptItemId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"FORBIDDEN",
				`Receipt "${receiptId}" cannot be updated while locked.`,
			);
		});
	});

	describe("functionality", () => {
		describe("update name", () => {
			runTests(() => ({ type: "name", name: faker.lorem.words() }));
		});

		describe("update price", () => {
			runTests(() => ({
				type: "price",
				price: Number(faker.finance.amount()),
			}));
		});

		describe("update quantity", () => {
			runTests(() => ({
				type: "quantity",
				quantity: faker.number.int({ max: 100 }),
			}));
		});

		describe("update locked", () => {
			runTests(() => ({ type: "locked", locked: true }), true);
		});
	});
});
