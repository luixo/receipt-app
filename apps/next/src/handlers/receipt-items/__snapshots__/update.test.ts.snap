// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`receiptItems.update > functionality > update locked > foreign 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'Shoes', 1167.8300, 2.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', false);
+ INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'Shoes', 1167.8300, 2.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', true);"
`;

exports[`receiptItems.update > functionality > update locked > own 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'Hat', 8124.3700, 1.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', false);
@@ --- --- @@
+ INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'Hat', 8124.3700, 1.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', true);"
`;

exports[`receiptItems.update > functionality > update name > foreign 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'Shoes', 1167.8300, 2.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', false);
+ INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'repellat asperiores rerum', 1167.8300, 2.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', false);"
`;

exports[`receiptItems.update > functionality > update name > own 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'Hat', 8124.3700, 1.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', false);
@@ --- --- @@
+ INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'eos ad deleniti', 8124.3700, 1.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', false);"
`;

exports[`receiptItems.update > functionality > update price > foreign 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'Shoes', 1167.8300, 2.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', false);
+ INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'Shoes', 998.1600, 2.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', false);"
`;

exports[`receiptItems.update > functionality > update price > own 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'Hat', 8124.3700, 1.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', false);
@@ --- --- @@
+ INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'Hat', 180.0700, 1.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', false);"
`;

exports[`receiptItems.update > functionality > update quantity > foreign 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'Shoes', 1167.8300, 2.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', false);
+ INSERT INTO public.\\"receiptItems\\" VALUES ('f6ddac04-d233-4e5a-926c-680d1eeae1d5', 'Shoes', 1167.8300, 100.0000, 'd3a2a718-f51b-4355-8777-b4615e935fa6', false);"
`;

exports[`receiptItems.update > functionality > update quantity > own 1`] = `
"Snapshot Diff:
- Before snapshot
+ After snapshot

@@ --- --- @@
- INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'Hat', 8124.3700, 1.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', false);
@@ --- --- @@
+ INSERT INTO public.\\"receiptItems\\" VALUES ('11dfa3bf-8270-4233-8644-efa2066d2e0b', 'Hat', 8124.3700, 18.0000, 'ec6c1495-54c6-4c11-b965-73b5a08b1049', false);"
`;
